In [5]:
# Framework for hypothesis testing
# 1. Setup Null and Alternate Hypothesis
# 2. Choose the test statistic
# 3. Select the Left vs Right vs Two-Tailed test, as per the hypothesis
# 4. Compute the P-Value
# 5. Compare the P-Value to the Significance Level (α) and conclude the test.

In [6]:
from scipy.stats import norm

def zTest(
    sampleMean,
    sampleSize,
    populationMean,
    populationStd,
    alternative="two-sided",
    alpha=None,
):
    SE = populationStd / (sampleSize**0.5)
    z = (sampleMean - populationMean) / SE

    # calculate p
    if alternative == "less":
        p = norm.cdf(z)
    if alternative == "greater":
        p = 1 - norm.cdf(z)
    if alternative == "two-sided":
        p = 2 * (1 - norm.cdf(abs(z)))

    zRejectionRange = None
    if alpha:
        # z critical
        if alternative == "less":
            zRejectionRange = (-float("inf"), norm.ppf(alpha))
        if alternative == "greater":
            zRejectionRange = (norm.ppf(1 - alpha), float("inf"))
        if alternative == "two-sided":
            zRejectionRange = (-norm.ppf(1 - alpha / 2), +norm.ppf(1 - alpha / 2))

    return p, z, zRejectionRange

Question 1

A country has a population average height of 65 inches with a standard deviation of 2.5. A person feels people from his state are shorter. He takes the average of 20 people and sees that it is 64.5.

At a 5% significance level (or 95% confidence level), can we conclude that people from his state are shorter, using the Z-test? What is the p-value?

In [7]:
alpha = 0.05

# Framework for hypothesis testing
# 1. Setup Null and Alternate Hypothesis
H0 = "Average population in his state = 65"
H1 = "Average population in his state < 65"

# 2. Choose the test statistic
test = zTest

# 3. Select the Left vs Right vs Two-Tailed test, as per the hypothesis
alternative = "less"

# 4. Compute the P-Value
p, z, zRejectionRange = test(64.5, 20, 65, 2.5, alternative, alpha)
print(f"p value = {round(p, 3)}")
print(f"zRejectionRange = {zRejectionRange}, z = {z}")

# 5. Compare the P-Value to the Significance Level (α) and conclude the test.
alpha = 0.05

if p < alpha:
    print(f"Reject: {H0}")
else:
    print(f"Failed to reject: {H0}")

p value = 0.186
zRejectionRange = (-inf, -1.6448536269514729), z = -0.8944271909999159
Failed to reject: Average population in his state = 65


Question 2

A French cafe has historically maintained that their average daily pastry production is at most 500.

With the installation of a new machine, they assert that the average daily pastry production has increased. The average number of pastries produced per day over a 70-day period was found to be 530.

Assume that the population standard deviation for the pastries produced per day is 125.

Perform a z-test with the critical z-value = 1.64 at the alpha (significance level) = 0.05 to evaluate if there's sufficient evidence to support their claim of the new machine producing more than 500 pastries daily.

Note: Round off the z-score to two decimal places.

In [8]:
alpha = 0.05

# Framework for hypothesis testing
# 1. Setup Null and Alternate Hypothesis
H0 = "Average daily pastry production = 500"
H1 = "Average daily pastry production > 500"

# 2. Choose the test statistic
test = zTest

# 3. Select the Left vs Right vs Two-Tailed test, as per the hypothesis
alternative = "greater"

# 4. Compute the P-Value
p, z, zRejectionRange = test(530, 70, 500, 125, alternative, alpha)
print(f"p value = {round(p, 3)}")
print(f"zRejectionRange = {zRejectionRange}, z = {z}")

# 5. Compare the P-Value to the Significance Level (α) and conclude the test.
alpha = 0.05

if p < alpha:
    print(f"Reject: {H0}")
else:
    print(f"Failed to reject: {H0}")

p value = 0.022
zRejectionRange = (1.6448536269514722, inf), z = 2.007984063681781
Reject: Average daily pastry production = 500


Question 3

The Chai Point stall at Bengaluru airport estimates that each person visiting the store drinks an average of 1.7 small cups of tea.

Assume a population standard deviation of 0.5 small cups. A sample of 30 customers collected over a few days averaged 1.85 small cups of tea per person.

Test the claim using an appropriate test at an alpha = 0.05 significance value, with a critical z-score value of ±1.96.

Note: Round off the z-score to two decimal places

In [9]:
alpha = 0.05

# Framework for hypothesis testing
# 1. Setup Null and Alternate Hypothesis
H0 = "Average number small cups of tea per person = 1.7"
H1 = "Average number small cups of tea per person != 1.7"

# 2. Choose the test statistic
test = zTest

# 3. Select the Left vs Right vs Two-Tailed test, as per the hypothesis
alternative = "two-sided"

# 4. Compute the P-Value
p, z, zRejectionRange = test(1.85, 30, 1.7, 0.5, alternative, alpha)
print(f"p value = {round(p, 3)}")
print(f"zRejectionRange = {zRejectionRange}, z = {z}")

# 5. Compare the P-Value to the Significance Level (α) and conclude the test.
alpha = 0.05

if p < alpha:
    print(f"Reject: {H0}")
else:
    print(f"Failed to reject: {H0}")

p value = 0.1
zRejectionRange = (-1.959963984540054, 1.959963984540054), z = 1.6431676725155
Failed to reject: Average number small cups of tea per person = 1.7


Question 4

A data scientist is looking at how a web application responds, with an average response time of 250 milliseconds and a standard deviation of 30 milliseconds.

Find the critical value for a 96% confidence level.

In [10]:
_, _, zRejectionRange = zTest(0, 1, 250, 30, alpha=(1-0.96))

from numpy import array
250 + array(zRejectionRange)*30

array([188.38753268, 311.61246732])

Question 5

A marketing team aims to estimate the average time, visitors spend on their website.

They gathered a random sample of 100 visitors and determined that the average time spent on the website was 4.5 minutes.

The team is working under the assumption that the population's mean time spent on the website is 4.0 minutes, with a standard deviation of 1.2 minutes.

Their goal is to estimate the true time spent on the website with a 95% confidence level. Calculate the confidence interval values and make a conclusion based on the calculated interval.

In [18]:
_, _, zRejectionRange = zTest(0, 1, 4, 1.2, alpha=(1-0.95))

from numpy import array
4 + array(zRejectionRange)*1.2/(100*0.5)

array([3.95296086, 4.04703914])